tests/CollectorTest.php

```diff
@@ -14,27 +14,566 @@
+		$c = new Collector(self::getNotOrderedArray());
+		$c->orderByDesc('type', false);
+
+		$this->assertEquals(5, $c->count());
+		$this->assertInstanceOf(PublicView::class, $c->get()[0]);
+		$this->assertSame(2, $c->get()[0]->get('type'));
+		$this->assertInstanceOf(PublicView::class, $c->get()[1]);
+		$this->assertSame(2, $c->get()[1]->get('type'));
+		$this->assertInstanceOf(PublicView::class, $c->get()[2]);
+		$this->assertSame(1, $c->get()[2]->get('type'));
+		$this->assertInstanceOf(PublicView::class, $c->get()[3]);
+		$this->assertSame(0, $c->get()[3]->get('type'));
+		$this->assertInstanceOf(PublicView::class, $c->get()[4]);
+		$this->assertSame(0, $c->get()[4]->get('type'));
+
+		//Test "OrderByDesc" integer case-sensitive
+		$c = new Collector(self::getNotOrderedArray());
+		$c->orderByDesc('type', true);
+
+		$this->assertEquals(5, $c->count());
+		$this->assertInstanceOf(PublicView::class, $c->get()[0]);
+		$this->assertSame(2, $c->get()[0]->get('type'));
+		$this->assertInstanceOf(PublicView::class, $c->get()[1]);
+		$this->assertSame(2, $c->get()[1]->get('type'));
+		$this->assertInstanceOf(PublicView::class, $c->get()[2]);
+		$this->assertSame(1, $c->get()[2]->get('type'));
+		$this->assertInstanceOf(PublicView::class, $c->get()[3]);
+		$this->assertSame(0, $c->get()[3]->get('type'));
+		$this->assertInstanceOf(PublicView::class, $c->get()[4]);
+		$this->assertSame(0, $c->get()[4]->get('type'));
+	}
+
+	public function test_where_has(): void {
+		//Test "WhereHas" returns data
+		$c = new Collector(self::getArray());
+		$c->whereHas('label');
+		$this->assertEquals(2, $c->count());
+		$this->assertInstanceOf(PublicView::class, $c->first());
+		$this->assertSame('first_label', $c->first()->get('label'));
+		$this->assertInstanceOf(PublicView::class, $c->last());
+		$this->assertSame('third_label', $c->last()->get('label'));
+
+		//Test "WhereHas" returns no data
+		$c = new Collector(self::getNotOrderedArray());
+		$c->whereHas('non-existant-key');
+		$this->assertEquals(0, $c->count());
+	}
+
+	public function test_where_has_not(): void {
+		//Test "WhereHasNot" returns data
+		$c = new Collector(self::getArray());
+		$c->whereHasNot('label');
+		$this->assertEquals(4, $c->count());
+
+		//Test "WhereHasNot" returns no data
+		$c = new Collector(self::getNotOrderedArray());
+		$c->whereHasNot('path');
+		$this->assertEquals(0, $c->count());
+	}
+
+	public function test_exclude(): void {
+		//Test "Exclude" called with object other than PublicView
+		$c = new Collector(self::getArray());
+		try {
+			$c->exclude([1, 2]);
+		} catch (\Exception $e) {
+			$this->assertEquals(\Exception::class, get_class($e));
+			$this->assertEquals('exclude() expects an array of PublicView objects.', $e->getMessage());
+		}
+
+		//Test "Exclude" called with one entry removes single entry
+		$c = new Collector(self::getArray());
+		$c->exclude(new PublicView(new View('first', View::TYPE_PAGE, 'first_label')));
```

Comment:
here you could just do `$c->exclude($c->first());` ... actually it's pretty weird it excludes another object that's not the same, i'll probably have to look into that.. but you can just leave it be